File: ScheduleWebApp/ClientApp/src/app/_services/lists.service.ts
import { Injectable } from '@angular/core';
import { Cabinet, Discipline, DisciplineType, Group, Teacher } from '../_models';
import { BehaviorSubject, Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../environments/environment';
import { map } from 'rxjs/operators';
import { Lists } from '../_models/lists';

@Injectable({
  providedIn: 'root'
})
export class ListsService {
  private disciplines: BehaviorSubject<Array<Discipline>> = new BehaviorSubject<Array<Discipline>>(new Array<Discipline>());
  private disciplineTypes: BehaviorSubject<Array<DisciplineType>> = new BehaviorSubject<Array<DisciplineType>>(new Array<DisciplineType>());
  private cabinets: BehaviorSubject<Array<Cabinet>> = new BehaviorSubject<Array<Cabinet>>(new Array<Cabinet>());
  private teachers: BehaviorSubject<Array<Teacher>> = new BehaviorSubject<Array<Teacher>>(new Array<Teacher>());
  private groups: BehaviorSubject<Array<Group>> = new BehaviorSubject<Array<Group>>(new Array<Group>());

  getDisciplines(): Observable<Array<Discipline>> {
    return this.disciplines.asObservable();
  }

  getDisciplineTypes(): Observable<Array<DisciplineType>> {
    return this.disciplineTypes.asObservable();
  }

  getCabinets(): Observable<Array<Cabinet>> {
    return this.cabinets.asObservable();
  }

  getTeachers(): Observable<Array<Teacher>> {
    return this.teachers.asObservable();
  }

  getGroups(): Observable<Array<Group>> {
    return this.groups.asObservable();
  }

  constructor(private http: HttpClient) {
  }

  addGroup(group: Group) {
    // TODO save group on server, get it and add to list
    const oldList = this.groups.value;
    oldList.push(group);
    this.groups.next(oldList);
  }

  addTeacher(teacher: Teacher) {
    // TODO save teacher on server, get it and add to list
    const oldList = this.teachers.value;
    oldList.push(teacher);
    this.teachers.next(oldList);
  }

  addDiscipline(discipline: Discipline) {
    // TODO save discipline on server, get it and add to list
    const oldList = this.disciplines.value;
    oldList.push(discipline);
    this.disciplines.next(oldList);
  }

  fetchForDirectorPage() {
    this.http.get<Lists>(`${environment.apiUrl}/lists`).subscribe(
      lists => {
        this.disciplines.next(lists.disciplines);
        this.disciplineTypes.next(lists.disciplineTypes);
        this.cabinets.next(lists.cabinets);
        this.teachers.next(lists.teachers);
        this.groups.next(lists.groups);
      });
  }
}

File: ScheduleWebApp/Services/ListsService.cs
using Database.Entities;
using ScheduleWebApp.Helpers;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ScheduleWebApp.Services
{
    public interface IListsService
    {
        List<Cabinet> GetCabinets();
        List<Discipline> GetDisciplines();
        List<DisciplineType> GetDisciplineTypes();
        List<Teacher> GetTeachers();
        List<StudyGroup> GetStudyGroups();
        void AddDiscipline(Discipline discipline);
        void AddDisciplineType(DisciplineType disciplineType);
        void AddCabinet(Cabinet cabinet);
        void AddTeacher(Teacher teacher);
        void AddGroup(StudyGroup group);
    }

    public class ListsService : IListsService
    {
        private DataContext _context;

        public ListsService(DataContext context)
        {
            _context = context;
        }

        public List<Cabinet> GetCabinets()
        {
            return _context.Cabinet.ToList();
        }

        public List<Discipline> GetDisciplines()
        {
            return _context.Discipline.ToList();
        }

        public List<DisciplineType> GetDisciplineTypes()
        {
            return _context.DisciplineType.ToList();
        }

        public List<Teacher> GetTeachers()
        {
            return _context.Teacher.ToList();
        }

        public void AddDiscipline(Discipline discipline)
        {
            _context.Discipline.Add(discipline);
            _context.SaveChanges();
        }

        public void AddDisciplineType(DisciplineType disciplineType)
        {
            _context.DisciplineType.Add(disciplineType);
            _context.SaveChanges();
        }

        public void AddCabinet(Cabinet cabinet)
        {
            _context.Cabinet.Add(cabinet);
            _context.SaveChanges();
        }

        public void AddTeacher(Teacher teacher)
        {
            _context.Teacher.Add(teacher);
            _context.SaveChanges();
        }

        public List<StudyGroup> GetStudyGroups()
        {
            return _context.StudyGroup.ToList();
        }

        public void AddGroup(StudyGroup group)
        {
            _context.StudyGroup.Add(group);
            _context.SaveChanges();
        }
    }
}
